core/templates/services/exploration-improvements-task-registrar.service.ts

```diff
@@ -0,0 +1,424 @@
+      stateAnswerStats: Map<string, AnswerStats[]>,
+      cstPlaythroughsByStateName: Map<string, CstPlaythrough[]>,
+      eqPlaythroughsByStateName: Map<string, EqPlaythrough[]>,
+      misPlaythroughsByStateName: Map<string, MisPlaythrough[]>) {
+    this.validateInitializeArgs(
+      expId, expVersion, states, expStats, openTasks,
+      resolvedTaskTypesByStateName, stateAnswerStats,
+      cstPlaythroughsByStateName, eqPlaythroughsByStateName,
+      misPlaythroughsByStateName);
+
+    this.expId = expId;
+    this.expVersion = expVersion;
+    this.expStats = expStats;
+
+    const allStateNames = states.getStateNames();
+    this.initializeStatsRegistry(
+      allStateNames, stateAnswerStats, cstPlaythroughsByStateName,
+      eqPlaythroughsByStateName, misPlaythroughsByStateName);
+    this.initializeTaskRegistry(
+      allStateNames, openTasks, resolvedTaskTypesByStateName);
+
+    for (const [stateName, taskRegistry] of this.taskRegistry.entries()) {
+      const statsRegistry = this.statsRegistry.get(stateName);
+      taskRegistry.forEach(t => this.refreshTask(t, statsRegistry));
+    }
+  }
+
+  flushInitializationChanges(): ITaskEntryPayloadDict[] {
+    if (this.initializationChangesAreFlushed) {
+      return [];
+    }
+    this.initializationChangesAreFlushed = true;
+
+    const payload = [];
+    for (const taskRegistry of this.taskRegistry.values()) {
+      if (taskRegistry.didHbrTaskStatusChange()) {
+        payload.push(taskRegistry.hbrTask.toPayloadDict());
+      }
+    }
+    return payload;
+  }
+
+  onStateAdd(stateName: string): void {
+    const newTasks = ImprovementsConstants.TASK_TYPES.map(
+      taskType => this.explorationTaskObjectFactory.createNewObsoleteTask(
+        this.expId, this.expVersion, taskType, stateName));
+    this.taskRegistry.set(stateName, new TaskRegistry(newTasks));
+    this.statsRegistry.set(stateName, new StatsRegistry());
+    newTasks.forEach(t => this.tasksByType.get(t.taskType).push(t));
+  }
+
+  onStateDelete(stateName: string): void {
+    const taskRegistry = this.taskRegistry.get(stateName);
+    taskRegistry.forEach(t => t.markAsObsolete());
+
+    // eslint-disable-next-line dot-notation
+    this.taskRegistry.delete(stateName);
+    // eslint-disable-next-line dot-notation
+    this.statsRegistry.delete(stateName);
+  }
+
+  onStateRename(oldStateName: string, newStateName: string): void {
+    const oldTaskRegistry = this.taskRegistry.get(oldStateName);
+    const newTaskRegistry = (
+      oldTaskRegistry.map(t => t.cloneWithNewTarget(newStateName)));
+
+    newTaskRegistry.forEach(t => this.tasksByType.get(t.taskType).push(t));
+    oldTaskRegistry.forEach(t => t.markAsObsolete());
+
+    // eslint-disable-next-line dot-notation
+    this.taskRegistry.delete(oldStateName);
+    // eslint-disable-next-line dot-notation
+    this.statsRegistry.delete(oldStateName);
+
+    this.taskRegistry.set(newStateName, newTaskRegistry);
```

Comment:
Goodness, no ... you cannot use "task registry" to simultaneously refer to a mapping and an item within that mapping. That leads to serious confusion. Please choose better names.